src/vulkan/frame_buffer.cc

```diff
@@ -156,24 +127,46 @@ Result FrameBuffer::ChangeFrameImageLayout(CommandBuffer* command,
           : VK_PIPELINE_STAGE_TRANSFER_BIT;
 
   for (auto& img : color_images_) {
-    img->ChangeLayout(command, old_layout,
-                      VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL, source_stage,
+    auto barrier = img->CreateBarrier(old_layout,
+                                      VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL);
+    img->ImageBarrier(command, barrier, source_stage,
                       VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT);
   }
   if (depth_image_) {
-    depth_image_->ChangeLayout(
-        command, old_layout, VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL,
-        source_stage, VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT);
+    auto barrier = depth_image_->CreateBarrier(
+        old_layout, VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL);
+    depth_image_->ImageBarrier(command, barrier, source_stage,
+                               VK_PIPELINE_STAGE_LATE_FRAGMENT_TESTS_BIT);
   }
   frame_image_layout_ = FrameImageState::kClearOrDraw;
-  return {};
 }
 
-void FrameBuffer::CopyColorImagesToHost(CommandBuffer* command) {
+void FrameBuffer::ChangeFrameToProbeLayout(CommandBuffer* command) {
+  if (frame_image_layout_ == FrameImageState::kProbe)
+    return;
+
   for (auto& img : color_images_) {
-    ChangeFrameImageLayout(command, FrameImageState::kProbe);
-    img->CopyToHost(command);
+    auto barrier = img->CreateBarrier(VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL,
+                                      VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL);
+    img->ImageBarrier(command, barrier,
+                      VK_PIPELINE_STAGE_COLOR_ATTACHMENT_OUTPUT_BIT,
+                      VK_PIPELINE_STAGE_TRANSFER_BIT);
```

Comment:
This code assumes ChangeFrameToProbeLayout() is called only when frame_image_layout is either kProbe or kClearOrDraw. While that may be the case now, wouldn't it be good and more future proof to somehow treat the kInit case (or any other cases)?

Reply:
This is an excellent point, and one I encountered 3 patch sets down. I have a CL coming which stores the current image layout in the image and allows doing any transition instead of just ones we thought of.